idl/storage_proxy.idl.hh

```diff
@@ -22,7 +23,13 @@
 #include "idl/keys.idl.hh"
 #include "idl/uuid.idl.hh"
 
-verb [[with_client_info, with_timeout, one_way]] mutation (frozen_mutation fm, inet_address_vector_replica_set forward, gms::inet_address reply_to, unsigned shard, uint64_t response_id, std::optional<tracing::trace_info> trace_info [[version 1.3.0]], db::per_partition_rate_limit::info rate_limit_info [[version 5.1.0]]);
+namespace service {
+    struct fencing_token {
+        int64_t topology_version;
+    };
+}
+
+verb [[with_client_info, with_timeout, one_way]] mutation (frozen_mutation fm, inet_address_vector_replica_set forward, gms::inet_address reply_to, unsigned shard, uint64_t response_id, std::optional<tracing::trace_info> trace_info [[version 1.3.0]], db::per_partition_rate_limit::info rate_limit_info [[version 5.1.0]], service::fencing_token fencing_token [[version 5.2.0]]);
```

Comment:
I don't think we need to include the token with every verb. It's enough to negotiate the token when the connection is first established (and if we decide to forcibly evict a member, tear down its connections).

Reply:
Do you propose to drop all connections each time a topology change happens although most verbs do not even care about it?

Reply:
For a stale node, why not?

If we're talking about a node using a stale topology, that's a different matter. We first have to agree what problem we're solving. $subject says "from stale nodes", but I think we should expand the scope (see alternative algorithm below).

Reply:
Stale node here is a node with a stale topology.

Reply:
We can't just reject nodes with stale topology based on version, it converts an rpc that started before the topology change reached the node to a failed rpc, even if the new topology is compatible with the old topology.

We need a smooth transition and only reject requests for incompatible topologies.

Reply:
We can if we retry the request based on the error (we may not need to if CL is satisfied even with some nodes returning the error). Anyway for a request with newer topology there is no need to answer with error, just wait for a new topology to propagate (raft::read_barrier()). And for a request with older topology the replica may verify that it is still a valid replica with newer topology for the request and return an error only if it is not.

Reply:
This just adds latency (or work for hints/repair if CL was satisfied).

There's no reason to reject valid reads/writes just because something happened elsewhere.

Reply:
If it is not in a replica set of the keyspace the check on "some node" against the topology will show that a token  still belongs to "some node" and it will request served.  What fencing will do is to force the check in case the request has older topology which is not usually done.

Reply:
I like the idea (fast-path and slow-path) but it really should have been explained up-front.

Should we introduce the slow-path first, then add the fast-path as an optimization?

Reply:
May be. What we are doing now it codding from two direction first is to move topology change to raft to get linearizable  epoch and second is to introduce  fencing that will use the epoch that does not exist yet. We will meet in the middle.